evap/static/js/datagrid.js

```diff
@@ -0,0 +1,314 @@
+        this.container = container;
+        this.sortableColumns = Array.from(this.sortableHeaders.keys());
+        this.searchInput = searchInput;
+        this.rows = this.fetchRowData();
+        this.loadFromStorage();
+        this.bindEvents();
+    }
+
+    bindEvents() {
+        this.delayTimer = null;
+        this.searchInput.on("change paste input", () => {
+            clearTimeout(this.delayTimer);
+            this.delayTimer = setTimeout(() => {
+                this.search(this.searchInput.val());
+            }, 200);
+        }).keypress(event => {
+            // after enter, unfocus the search input to collapse the screen keyboard
+            if (event.keyCode === 13) {
+                event.target.blur();
+            }
+        });
+
+        for (const [column, header] of this.sortableHeaders) {
+            header.click(() => {
+                const ordering = header.hasClass("col-sorting-asc") ? "desc" : "asc";
+                this.sort([[column, ordering]]);
+            });
+        }
+    }
+
+    fetchRowData() {
+        return this.container.children().get()
+            .map(row => {
+                let data = {
+                    element: row,
+                    orderValues: new Map(),
+                    filterValues: new Map(),
+                };
+                data.searchWords = this.findSearchableCells($(row))
+                    .flatMap(element => this.searchWordsOf($(element).text()));
+
+                this.addFilterRowData($(row), data);
+
+                for (const column of this.sortableColumns) {
+                    const cell = this.findCellByColumn($(row), column);
+                    if (cell.is("[data-order]")) {
+                        data.orderValues.set(column, cell.data("order"));
+                    } else {
+                        data.orderValues.set(column, cell.html().trim());
+                    }
+                }
+                return data;
+            });
+    }
+
+    searchWordsOf(string) {
+        return string.toLowerCase().trim().split(/\s+/);
+    }
+
+    search(searchValue) {
+        this.state.search = searchValue;
+        this.filterRows();
+        this.update();
+    }
+
+    filter(filter) {
+        Object.assign(this.state.filter, filter);
+        this.filterRows();
+        this.update();
+    }
+
+    // Filters rows respecting the current search string and filters by their searchWords and filterValues
+    filterRows() {
+        const searchWords = this.searchWordsOf(this.state.search);
+        for (const row of this.rows) {
+            const isDisplayedBySearch = searchWords.every(searchWord => {
+                return row.searchWords.some(rowWord => rowWord.includes(searchWord));
+            });
+            const isDisplayedByFilters = Object.entries(this.state.filter).every(([name, filterValues]) => {
+                if (filterValues.length === 0) {
+                    return true;
+                } else {
```

Comment:
because of the return i'd remove the else and de-indent the code below

Reply:
Resolved as filters now are never empty. This moves this logic away from `filterRows`. Also, having only "active" filters in the map is imo a simpler mental model.